src/theory/strings/core_solver.cpp

```diff
@@ -2286,29 +2291,27 @@ void CoreSolver::checkLengthsEqc() {
   }
 }
 
-void CoreSolver::doInferInfo(const InferInfo& ii)
+bool CoreSolver::processInferInfo(CoreInferInfo& ii)
 {
-  // send the inference
+  // rewrite the conclusion, ensure non-trivial
+  ii.d_conc = Rewriter::rewrite(ii.d_conc);
+
+  if (ii.isTrivial())
+  {
+    // rewrote to true
+    return false;
+  }
+  // process the state change to this solver
   if (!ii.d_nf_pair[0].isNull())
   {
     Assert(!ii.d_nf_pair[1].isNull());
     addNormalFormPair(ii.d_nf_pair[0], ii.d_nf_pair[1]);
   }
-  // send the inference
-  d_im.sendInference(ii);
-  // Register the new skolems from this inference. We register them here
-  // (lazily), since the code above has now decided to use the inference
-  // at use_index that involves them.
-  for (const std::pair<const LengthStatus, std::vector<Node> >& sks :
-       ii.d_new_skolem)
-  {
-    for (const Node& n : sks.second)
-    {
-      d_termReg.registerTermAtomic(n, sks.first);
-    }
-  }
-}
+  // send the inference, which is a lemma
+  d_im.sendInference(ii, true);
```

Comment:
Note that `ii` will be sliced here. If we intend to have virtual methods in `InferInfo`, this will not work and we will instead have to pass a pointer. We could make the vectors holding inferences/lemmas `std::vector<std::unique_ptr<InferInfo>>`.

Reply:
I think the solution (which I had an earlier version of this PR) is to make CoreInferInfo a wrapper of InferInfo.

Generally, I think we should keep this core data structure simple. After some thought, I think I would not like virtual methods in InferInfo.

Reply:
Yeah, I think I like this approach.